napari/_qt/_qapp_model/qactions/_file.py

```diff
@@ -0,0 +1,169 @@
+    ),
+    Action(
+        id=CommandId.DLG_OPEN_FILES_AS_STACK,
+        title=CommandId.DLG_OPEN_FILES_AS_STACK.title,
+        callback=QtViewer._open_files_dialog_as_stack_dialog,
+        menus=[{'id': MenuId.MENUBAR_FILE, 'group': MenuGroup.NAVIGATION}],
+        keybindings=[{'primary': KeyMod.CtrlCmd | KeyMod.Alt | KeyCode.KeyO}],
+    ),
+    Action(
+        id=CommandId.DLG_OPEN_FOLDER,
+        title=CommandId.DLG_OPEN_FOLDER.title,
+        callback=QtViewer._open_files_dialog,
+        menus=[{'id': MenuId.MENUBAR_FILE, 'group': MenuGroup.NAVIGATION}],
+        keybindings=[
+            {'primary': KeyMod.CtrlCmd | KeyMod.Shift | KeyCode.KeyO}
+        ],
+    ),
+    Action(
+        id=CommandId.DLG_OPEN_FILES_WITH_PLUGIN,
+        title=CommandId.DLG_OPEN_FILES_WITH_PLUGIN.title,
+        callback=_open_files_with_plugin,
+        menus=[
+            {'id': MenuId.FILE_OPEN_WITH_PLUGIN, 'group': MenuGroup.NAVIGATION}
+        ],
+    ),
+    Action(
+        id=CommandId.DLG_OPEN_FILES_AS_STACK_WITH_PLUGIN,
+        title=CommandId.DLG_OPEN_FILES_AS_STACK_WITH_PLUGIN.title,
+        callback=_open_files_as_stack_with_plugin,
+        menus=[
+            {'id': MenuId.FILE_OPEN_WITH_PLUGIN, 'group': MenuGroup.NAVIGATION}
+        ],
+    ),
+    Action(
+        id=CommandId.DLG_OPEN_FOLDER_WITH_PLUGIN,
+        title=CommandId.DLG_OPEN_FOLDER_WITH_PLUGIN.title,
+        callback=_open_folder_with_plugin,
+        menus=[
+            {'id': MenuId.FILE_OPEN_WITH_PLUGIN, 'group': MenuGroup.NAVIGATION}
+        ],
+    ),
+    Action(
+        id=CommandId.DLG_SHOW_PREFERENCES,
+        title=CommandId.DLG_SHOW_PREFERENCES.title,
+        callback=Window._open_preferences_dialog,
+        menus=[{'id': MenuId.MENUBAR_FILE, 'group': MenuGroup.PREFERENCES}],
+        # TODO: revert to `StandardKeyBinding.Preferences` after app-model>0.2.0
+        keybindings=[{'primary': KeyMod.CtrlCmd | KeyCode.Comma}],
+    ),
+    # TODO!
+    # this action would conflict because having the same id as below... but
+    # we could check whether the args are the same and if not, then allow registration
+    # Action(
+    #     id=CommandId.DLG_SAVE_LAYERS,
+    #     title=CommandId.DLG_SAVE_LAYERS.title,
+    #     callback=QtViewer._save_layers_dialog,
+    #     kwargs={'selected': True},  # <<<<< TODO!
+    #     menus=[{'id': MenuId.MENUBAR_FILE, 'group': '3_'}],
+    #     keybindings=[StandardKeyBinding.Save],
+    #     enablement='num_selected_layers > 0'
+    # ),
+    Action(
+        id=CommandId.DLG_SAVE_LAYERS,
+        title=trans._('Save All Layers...'),
+        callback=QtViewer._save_layers_dialog,
+        menus=[{'id': MenuId.MENUBAR_FILE, 'group': MenuGroup.SAVE}],
+        # Conflict: ctrl+shift+s typically means "save as ..."
+        keybindings=[
+            {'primary': KeyMod.CtrlCmd | KeyMod.Shift | KeyCode.KeyS}
+        ],
+        enablement=(LLCK.num_selected_layers > 1),
```

Comment:
This does not work, probably the reason Talley had this as a string originally. Menubar items are built via:

In `Window._add_menus`, `build_qmodel_menu(<MenuID>)` is run -> inits `QModelMenu` on the `MenuID`. During init, runs [`_rebuild`](https://github.com/pyapp-kit/app-model/blob/5bc0300d9025389130ef6dacc20ef12aa93c8dbf/src/app_model/backends/qt/_qmenu.py#L307) (in app-model), which inits [`QMenuItemAction`](https://github.com/pyapp-kit/app-model/blob/5bc0300d9025389130ef6dacc20ef12aa93c8dbf/src/app_model/backends/qt/_qaction.py#L112), which unfortunately runs:

```python
        # by updating from an empty context, anything that declares a "constant"
        # enablement expression (like `'False'`) will be evaluated, allowing any
        # menus that are always on/off, to be shown/hidden as needed.
        # Everything else will fail without a proper context.
        # TODO: as we improve where the context comes from, this could be removed.
        with contextlib.suppress(NameError):
            self.update_from_context({})
```

Because context is empty, the enablement [expression](https://app-model.readthedocs.io/en/latest/expressions/) above always evaluates to `True`. Not sure exactly what is meant by 'Everything else will fail without a proper context.' - no error gets raised, just silently doesn't do what you wanted.

I am a little confused by the first sentence. A constant enablement expression (e.g., `True`) will evaluate correctly with a non-empty context. E.g.,

```
a = Expr.parse('True')
a.eval({'a':1, 'b':2})
True
```
 Or does it mean that only constant enablement expressions should be evaluated here?

Reply:
The comment is not suggesting that constant expressions are not able to be evaluated with a non-empty context (because they are), but, as the last line alludes to, app-model currently does not handle getting the context. You must deal with updating from context yourself, e.g., the layer menu is updated after building the menu:

https://github.com/napari/napari/blob/f2d7d437024c8e417687ffebee5fa2bc842d7941/napari/_qt/containers/_layer_delegate.py#L195-L206

For menubar menus, which always appear, we must connect an event to an update function so it is updated before the menu is shown.